frontend/src/metabase/visualizations/components/settings/ChartSettingFieldsPartition.jsx

```diff
@@ -44,148 +31,121 @@ class ChartSettingFieldsPartition extends React.Component {
-      this.setState({ displayedValue: null });
-    }
+
+  getPartitionType = partitionName => {
+    return partitionName === "rows" || partitionName === "columns"
+      ? "dimension"
+      : "metric";
   };
 
-  render() {
-    const value = _.mapObject(
-      this.state.displayedValue || this.props.value || {},
-      fieldRefs =>
-        fieldRefs
-          .map(field_ref =>
-            this.props.columns.find(col => _.isEqual(col.field_ref, field_ref)),
-          )
-          .filter(col => col != null),
-    );
-    return (
-      <div>
-        {this.props.partitions.map(({ name, title, columnFilter }, index) => (
-          <Partition
-            key={index}
-            className={cx("py2", { "border-top": index > 0 })}
-            title={title}
-            columnFilter={columnFilter}
-            partitionName={name}
-            columns={value[name]}
-            value={value}
-            getColumnSettingValue={this.getColumnSettingValue}
-            onChangeColumnSetting={this.handleChangeColumnSetting}
-            onEditFormatting={this.handleEditFormatting}
-            updateDisplayedValue={this.updateDisplayedValue}
-            commitDisplayedValue={this.commitDisplayedValue}
-          />
-        ))}
-      </div>
-    );
-  }
-}
+  handleDragEnd = ({ source, destination }) => {
+    if (!source || !destination) {
+      return;
+    }
+    const { value, onChange } = this.props;
+    const { droppableId: sourcePartition, index: sourceIndex } = source;
+    const { droppableId: destinationPartition, index: destinationIndex } =
+      destination;
 
-class PartitionInner extends React.Component {
-  render() {
-    const {
-      columns = [],
-      partitionName,
-      columnFilter,
-      onChangeColumnSetting,
-      getColumnSettingValue,
-      onEditFormatting,
-      updateDisplayedValue,
-      commitDisplayedValue,
-      value,
-      connectDropTarget,
-      title,
-      className,
-    } = this.props;
-    return connectDropTarget(
-      <div className={className}>
-        <Label color="medium">{title}</Label>
-        {columns.length === 0 ? (
-          <EmptyPartition
-            columnFilter={columnFilter}
-            value={value}
-            partitionName={partitionName}
-            updateDisplayedValue={updateDisplayedValue}
-          />
-        ) : (
-          columns.map((col, index) => (
-            <Column
-              key={index}
-              partitionName={partitionName}
-              column={col}
-              index={index}
-              columnFilter={columnFilter}
-              value={value}
-              onChangeColumnSetting={onChangeColumnSetting}
-              getColumnSettingValue={getColumnSettingValue}
-              onEditFormatting={onEditFormatting}
-              updateDisplayedValue={updateDisplayedValue}
-              commitDisplayedValue={commitDisplayedValue}
-            />
-          ))
-        )}
-      </div>,
-    );
-  }
-}
+    if (
```

Comment:
This logic is kinda gross, but it's what was originally here. I'd be curious if folks can think of any improvements. Being a map of arrays, I suspect shallow cloning the map gets a little awkward.